### PR TITLE
enable advanced scroll&pan customizations

### PR DESCRIPTION
**Pull request by vonovak**
## Motivation

Hello and thanks for the lib! The motivation for this change is that some behaviors of the bottom sheet are hard-wired into the implementation and I'd like them to be customizable. Specifically, I needed to implement a custom behavior of the bottom sheet that is visible in this recording:

https://user-images.githubusercontent.com/1566403/124756778-2445ab00-df2d-11eb-9088-00c1ed1aa3f3.mp4

(if you're wondering how the sheet gets back down, it's by dragging or tapping the handle)

I tried to make the changes so that:

- they do not hamper maintainability of existing code and do not make your life as maintainer harder
- adding complex behaviors is not easy but is possible: it requires a little more coding on the consumer side. Basically, some of the code that is in the library and offers default behavior, needs to be copied over to the userland and the necessary modifications are done in it.

The changes necessary to implement the custom behavior are rather small and are as follows:

1) I moved handleGestureStart, handleGestureActive and handleGestureEnd from `BottomSheet.tsx` into a separate hook, and that hook can be customized via a prop passed to `BottomSheet`
2) added `animatedSnapPoints` to be provided by `BottomSheetInternalContext`
3) added `scrollEventListenersHook` prop to Scrollables. It defaults to `useScrollableInternal`. Overriding this prop, again, enables advanced customizations

in userland:

1) create`useCustomScrollableInternal.tsx` (that is similar in function to `useScrollableInternal`) and `useCustomPanGestureHandlerListeners.ts`. These files are where the magic happens.
2) pass custom `panGestureHandlerListenersHook` prop to `BottomSheet` and `scrollEventListenersHook` to `BottomSheetScrollView` (or other Scrollables)

OR instead, we can create `useCustomScrollableInternal.tsx`, use it in `CreateCustomBottomSheetScrollableComponent.tsx` (similar to `CreateBottomSheetScrollableComponent`) and use that in `BottomSheetCustomScrollView.tsx` (smilar to `BottomSheetScrollView.tsx`) and then render `BottomSheetCustomScrollView` instead of `BottomSheetScrollView`. This is a lot more work in userland, but it means one fewer prop to support on the library side. You can see this approach in the [here](https://github.com/vonovak/react-native-bottom-sheet/tree/0521231b8a62b67f56f73384f1f27897d62e2c7d/example/src/screens/advanced/customScrollAndDrag); let me know what you prefer.

I added this as a demo to the example app as well, but I'm not sure if you want to include it because there is some code that is very similar to the code that already lives inside the bottom sheet implementation, and if you make changes to `useInteractivePanGestureHandlerListeners.ts` (new file added in this PR), you'd need to make similar changes to the example app as well. So let me know if you want to exclude the changes I did in the example app.

To be decided: 
1) naming of the new props
2) documentation. Personally, I think it's okay to flag these new props as experimental and subject to change even without a major version bump, because I see the parts are under active development, but I'll leave it up to you

appreciate your feedback on this one, thank you!  🙂

### CODE REVIEW

File: src/components/bottomSheet/BottomSheet.tsx [hunk 1]
```diff
@@ -948,341 +946,25 @@ const BottomSheetComponent = forwardRef<BottomSheet, BottomSheetProps>(
     }));
     //#endregion
 
-    //#region gesture methods
-    const handleGestureStart = useWorkletCallback(
```
Comment (vonovak) [inline src/components/bottomSheet/BottomSheet.tsx:961]: I just moved all of this over to src/components/bottomSheet/useInteractivePanGestureHandlerListeners.ts

**Comment by gorhom** [comment]
hi @vonovak , thanks for submitting this amazing pr ! i will review and merge it this weekend 👏

**Comment by gorhom** [comment]
hi @vonovak ,  thanks for this pr, it will unlock more custom advance usage to this library.

I have looked at the pr, and i am impressed by how well you understand this library and writing a very neat readable code 🔥. As you mentioned, this feature doesn't conflict or complicate the current code-base, and I am very looking forward to get it merge before the release of `v4` so we could give it more testing and improvements.

The pr looks fine, but here are some suggestions we could do we before we merge it:
 
1. instead passing all needed params to `usePanGestureHandlerListeners`, why not just `useBottomSheetInternal` hook to extract them. This will require us to create another context under the `BottomSheetInternalProvider`.
2. extract `translationY` from `useInteractivePanGestureHandler` for each gesture source and add it to `BottomSheetInternalContext`, this will make it easier to access gesture variables without the need to add custom solution like the one in examples.
3. rename props: `usePanGestureHandlerListeners` to ` panGestureHandlersHook`, `useScrollEventListeners` to ` scrollEventHandlersHook`.
4. rename `useScrollableInternal` to `useScrollEventHandlerDefault`
5. rename `useInteractivePanGestureHandlerListeners` to `usePanGestureHandlersDefault`
6. extract `handleSettingScrollable` from `useScrollableInternal` into a separate file or just add it into `createBottomSheetScrollableComponent`

Looking forward to hearing from you, and i am happy to jump and help completing this PR anytime.

thanks

**Comment by nightwind2020** [comment]
Hi guys, I would like to say some thoughts.
The other day, I was looking at RNUI's code and felt that his abstraction of gestures seemed more reusable. ( Didn't consider the cost of code refactoring, but I hope this information is useful.

https://github.com/wix/react-native-ui-lib/tree/master/src/components/panningViews
https://wix.github.io/react-native-ui-lib/docs/PanDismissibleView/

**Comment by vonovak** [comment]
hello @gorhom! Thanks for the feedback. Here are my answers: :)

1.  we cannot use `useBottomSheetInternal` here because it would be used outside of the BottomSheet where the BottomSheetInternalProvider is rendered.

I guess the code could be restructured to allow this but the changes required would be far greater, so I'm a little "scared" to go that route and find it easier to just pass the params directly, albeit I agree there are many 😄 

2. I also thought of adding `translationY` to the internal context, but when we do this, other people will come and say "I need translationX", "I need velocity", "I need X, Y and Z" and the internal context will grow bigger and bigger and the API surface of the library will grow, even though the library itself does not use internally `translationY` anywhere.

I'm afraid it could make maintenance harder in the long run for not so much benefit. And the alternative it fairly simple - have consumers handle it. After all, all I needed was to create a `useSharedValue` and a context that makes it available to its children to consume.

3. this one I'm not totally sure about - somehow I take it as a convention in the React world that hooks start with the `use` prefix, and I feel like it makes sense to keep it (but I'll rename them if you really like those names :))

4, 5, 6: good points, did these changes.

Let me know if there's anything else, and feel free to jump in and modify the code (I ticked "Allow edits by maintainers") - it might be faster than exchanging messages :)
What I did not address is exporting the newly created hooks, so they are not part of the `@gorhom/bottom-sheet` export, and I wrote no docs. I'm happy to write these later too - preferably would like to get this merged sooner rather than later because the v4 branch is a moving target and every commit there increases the chance of conflicts :)

Thanks again!

**Comment by gorhom** [comment]
Hi @vonovak 👋

> I guess the code could be restructured to allow this but the changes required would be far greater, so I'm a little "scared" to go that route and find it easier to just pass the params directly, albeit I agree there are many 😄

Yeah, this will require us to create a new context under `BottomSheetInternal` to be able to consume internal variables, I will pick this up 🙋‍♂️

> I also thought of adding translationY to the internal context, but when we do this, other people will come and say "I need translationX", "I need velocity", "I need X, Y and Z" and the internal context will grow bigger and bigger and the API surface of the library will grow, even though the library itself does not use internally translationY anywhere.

I agree, tho when i saw your implementation, i thought why not simplify it. however we could add it later if needed 👍

>this one I'm not totally sure about - somehow I take it as a convention in the React world that hooks start with the use prefix, and I feel like it makes sense to keep it (but I'll rename them if you really like those names :))

yeah i know, but i thought of it before for the hook prop `focusHook`, the name `useFocusHook` - for me at least - is more of an action rather than a name of an action.

>Let me know if there's anything else, and feel free to jump in and modify the code (I ticked "Allow edits by maintainers") - it might be faster than exchanging messages :)
What I did not address is exporting the newly created hooks, so they are not part of the @gorhom/bottom-sheet export, and I wrote no docs. I'm happy to write these later too - preferably would like to get this merged sooner rather than later because the v4 branch is a moving target and every commit there increases the chance of conflicts :)

Thanks, I will work on getting point 1 done tonight and get it merge too 👏

Thanks again for your amazing contributions!

**Comment by gorhom** [comment]
@vonovak sorry for the delay. I have refactored the gestures to be provided internally in a context, this allow custom provided gestures to use `useBottomSheetInternal` 👏

i did a bit of refactoring here and there, everything seemed working fine, however we have a problem with `animatedScrollableState`, it is currently a derived value based on the `animatedSheetState` value.

we could tweak this somehow by adding a new override variable from provided `useScrollEventsHandlers` hook.

let me know what do you think about this!

**Comment by vonovak** [comment]
hello @gorhom! :)

>  I have refactored the gestures to be provided internally in a context, this allow custom provided gestures to use useBottomSheetInternal

that'll make it nicer to consume 👍 

> however we have a problem with animatedScrollableState, it is currently a derived value based on the animatedSheetState value.

honestly, I'm not following exactly what the problem is; there have been more than a few changes from when I pushed the last time. Can you point me to the relevant places in code? Thanks! :)

**Comment by gorhom** [comment]
Hi @vonovak sorry i did not clarifying it enough, we are using `animatedScrollableState` to handle some internal functionalities and props like `decelerationRate` & `showsVerticalScrollIndicator `.

https://github.com/gorhom/react-native-bottom-sheet/blob/d77b3f17d85a707ffe20fab27d20fb40f92ff1b7/src/hooks/useScrollableInternal.ts#L47-L52

The `animatedScrollableState` depends on `animatedSheetState` and `animatedKeyboardState`, but since we override the functionality , we have to override this logic to set the scrollable state to `UNLOCKED`.
https://github.com/gorhom/react-native-bottom-sheet/blob/d77b3f17d85a707ffe20fab27d20fb40f92ff1b7/src/components/bottomSheet/BottomSheet.tsx#L360-L388

In https://github.com/gorhom/react-native-bottom-sheet/pull/525/commits/1cdd9cf37e4c41a3600b490beb8b6fff7082a4fc, I have this override functionality and updated your custom hook.

I think this PR ready, we should test it more and get it merge soon

**Comment by vonovak** [comment]
okay, let me test it out and report if it works as expected 👍

**Comment by vonovak** [comment]
hello, there is one issue present with https://github.com/gorhom/react-native-bottom-sheet/pull/525/commits/1cdd9cf37e4c41a3600b490beb8b6fff7082a4fc and that is that previously, consumer was allowed control over `scrollableAnimatedProps` returned from `useScrollEventHandlerDefault`:

https://github.com/vonovak/react-native-bottom-sheet/blob/bfebeca6c9207e4e612932a680e42ae5f5ac25cf/src/hooks/useScrollEventHandlerDefault.ts#L145

now this is no longer customizable and is given in 

https://github.com/vonovak/react-native-bottom-sheet/blob/1cdd9cf37e4c41a3600b490beb8b6fff7082a4fc/src/components/bottomSheetScrollable/createBottomSheetScrollableComponent.tsx#L56

and the provided `decelerationRate` makes scrolling up in the "custom gesture handling" example feel weird. Can we return the control over this to the hook, or should consumers have their own implementation of `createBottomSheetScrollableComponent`? (the first option seems not hard to do)

**Comment by gorhom** [comment]
Hi @vonovak,  the `decelerationRate` and `showsVerticalScrollIndicator` depends on the `animatedScrollableState` which can be overridden by `animatedScrollableOverrideState` 

https://github.com/vonovak/react-native-bottom-sheet/blob/1cdd9cf37e4c41a3600b490beb8b6fff7082a4fc/example/src/screens/advanced/customGestureHandling/useCustomScrollEventsHandlers.tsx#L88-L95

I extracted the `animatedProps` from `scrollEventsHandlersHook` because it break the single responsibility principle and would make it confusing for people. We could allow user to provide `animatedPropsHook` that could provide any animated props to the scrollable, what do you think ?

**Comment by vonovak** [comment]
Hey @gorhom okay makes sense. I figured out the change necessary to make the deceleration rate work the way I want, I'll push it shortly.

As for animatedPropsHook, I don't really need it, so I'd wrap it up and not add make the much PR bigger. When someone needs that functionality, it can be added later... 🙂

edit: it's going to take a little longer, https://github.com/gorhom/react-native-bottom-sheet/pull/525/commits/c23910c70fa8beb395dd6c7c15487bf0950bdea1 does not fix it

**Comment by vonovak** [comment]
hello @gorhom I will indeed need to override the `animatedProps` it seems. I thought about the custom `animatedPropsHook` idea and it seems that it perhaps shouldn't be necessary: I could pass the `animatedProps` here directly: https://github.com/vonovak/react-native-bottom-sheet/blob/0e7719cf0c988d572f7ed900bb7b2e7005f769ec/example/src/screens/advanced/customGestureHandling/CustomGestureHandling.tsx#L79

but that would require a change in how props are passed here: https://github.com/vonovak/react-native-bottom-sheet/blob/0e7719cf0c988d572f7ed900bb7b2e7005f769ec/src/components/bottomSheetScrollable/createBottomSheetScrollableComponent.tsx#L111 (and a few more places around).

Because it seemed fairly simple to implement, I went ahead and did it in https://github.com/gorhom/react-native-bottom-sheet/pull/525/commits/e25aad0124228dc2ce02b2350459f876e02db1cb

It resulted into seemingly weird order of props passed here https://github.com/vonovak/react-native-bottom-sheet/blob/18d2b7d2d0ccbeb118015cd3601d83525c967853/src/components/bottomSheetScrollable/createBottomSheetScrollableComponent.tsx#L99

but it actually makes sense - in the snippet: `overScrollMode`, `keyboardDismissMode` and `style` props can be passed to the component as seen here https://github.com/vonovak/react-native-bottom-sheet/blob/0b63dc3b45291498b634288335ed773b2199175e/src/components/bottomSheetScrollable/createBottomSheetScrollableComponent.tsx#L30. `scrollEventThrottle` can not be passed to the component because it is forbidden in the TS typings https://github.com/vonovak/react-native-bottom-sheet/blob/18d2b7d2d0ccbeb118015cd3601d83525c967853/src/components/bottomSheetScrollable/types.d.ts#L45.

That leaves `animatedProps` with a default implementation that can be overridden, as well as other scrollview props supported.


```
<ScrollableComponent
  animatedProps={scrollableAnimatedProps}
  {...rest}
  scrollEventThrottle={16}
  ref={scrollableRef}
  overScrollMode={overScrollMode}
  keyboardDismissMode={keyboardDismissMode}
  onScroll={scrollHandler}
  style={containerStyle}
/>
```

Let me know what you think about this approach! Thanks :)

tested the example app on both platforms and the example works as expected. Tried other examples as well, they work fine.

**Comment by gorhom** [comment]
Thanks @vonovak for the update, i think your approach is fine but we will need to mention it in the docs.

I am going to review this PR one more time and will merge it today.

Thanks alot!

**Comment by ozasadnyy** [comment]
> Thanks @vonovak for the update, i think your approach is fine but we will need to mention it in the docs.

@gorhom maybe you forgot, but there is still no documentation about this property :)

**Comment by gorhom** [comment]
@ozasadnyy i kept it out of the documentation because it is still kinda experimental and its api might change any time, but i will add it to my to-do list

**Comment by SipanP** [comment]
@gorhom Any chance making the declaration rate a prop for the Scrollable components?

**Pull request merged**